CPP20.md

```diff
@@ -0,0 +1,271 @@
+```c++
+for (T thing = f(); auto& x : thing.items()) {
+    mutate(&x);
+    log(x);
+}
+
+for (std::size_t i = 0; const auto& x : foo()) {
+    bar(x, i++);
+}
+```
+
+## C++20 Library Features
+
+### Concepts library
+The concepts library provides definitions of fundamental library concepts that can be used to perform
+compile-time validation of template arguments and perform function dispatch based on properties of types.
+
+**Core language concepts**
+- Same - specifies two types are the same
+- DerivedFrom - specifies that a type is derived from another type
+- ConvertibleTo - specifies that a type is implicitly convertible to another type
+- Common - specifies that two types share a common type
+- Integral - specifies that a type is an integral type
+
+**Comparison concepts**
+- Boolean - specifies that a type can be used in Boolean contexts
+- EqualityComparable - specifies that operator == is an equivalence relation
+
+**Object concepts**
+- Movable - specifies that an object of a type can be moved and swapped
+- Copyable - pecifies that an object of a type can be copied, moved, and swapped
+
+**Callable concepts**
+- Invocable - specifies that a callable type can be invoked with a given set of argument types
+- Predicate - specifies that a callable type is a Boolean predicate
+
+### Ranges library
+Ranges are an extension of the Standard Template Library and an abstration layer on top of iterators
+that makes its iterators and algorithms more powerful by making them composable.
+
+Filter a container using a predicate and transform it.
+```c++
+using namespace std::ranges;
+std::vector<int> vi {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
+auto rng = vi | view::remove_if([](int i){ return i % 2 == 1; })
+              | view::transform([](int i){ return std::to_string(i); });
+// rng == {"2", "4", "6", "8", "10"};
+```
+
+Generate an infinite list of integers starting at 1, square them, take the first 10, and sum them.
+```c++
+using namespace std::ranges;
+int sum = accumulate(view::ints(1)
+              | view::transform([](int i){ return i * i; })
+              | view::take(10), 0);
+```
+
+Generate a sequence on the fly with a range comprehension and initialize a vector with it.
+```c++
+using namespace std::ranges;
+std::vector<int> vi = view::for_each(view::ints(1, 10), [](int i){
+        return yield_from(view::repeat_n(i, i));
+    });
+// vi == {1, 2, 2, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5, 5,...}
+```
+
+Read data into a vector, sort it, and make it unique.
+```c++
+extern std::vector<int> read_data();
+using namespace std::ranges;
+std::vector<int> vi = read_data() | action::sort | action::unique;
+vi |= action::sort | action::unique; // Mutate the container in-place
+```
+
+### std::bind_front
+Generates a forwarding call wrapper for a callable object f with parameters `Args&&... args`.
+Calling this wrapper is equivalent to invoking f with its first sizeof...(Args) parameters bound to args.
```

Comment:
Calling this wrapper is equivalent to invoking `f` with its first `sizeof...(Args)` parameters bound to `args`.